package/kedro_viz/api/apps.py

```diff
@@ -44,32 +46,75 @@
 
 
+def _create_etag() -> str:
+    """Generate the current timestamp to use as etag."""
+    return str(time.time())
+
+
 def _create_base_api_app() -> FastAPI:
     return FastAPI(
         title="Kedro-Viz API", description="REST API for Kedro Viz", version=__version__
     )
 
 
-def create_api_app_from_project(project_path: Path) -> FastAPI:
-    """Create an API from a real Kedro project by adding the router to the FastAPI app."""
+def create_api_app_from_project(
+    project_path: Path, autoreload: bool = False
+) -> FastAPI:
+    """Create an API from a real Kedro project by adding the router to the FastAPI app.
+
+    Args:
+        project_path: Path to the Kedro project
+        autoreload: Whether the app should autoreload based on content change
+            in the Kedro project
+    Returns:
+        The FastAPI app
+    """
     app = _create_base_api_app()
     app.include_router(router)
     app.mount("/static", StaticFiles(directory=_HTML_DIR / "static"), name="static")
 
+    # everytime the server reloads, a new app with a new timestamp will be created.
+    # this is used as an etag embedded in the frontend for client to use when making requests.
+    app_etag = _create_etag()
+
     @app.get("/")
     async def index():
         heap_app_id = kedro_telemetry.get_heap_app_id(project_path)
         heap_user_identity = kedro_telemetry.get_heap_identity()
         should_add_telemetry = bool(heap_app_id) and bool(heap_user_identity)
         html_content = (_HTML_DIR / "index.html").read_text(encoding="utf-8")
+        injected_head_content = []
+
+        env = Environment(loader=FileSystemLoader(_HTML_DIR))
         if should_add_telemetry:
-            env = Environment(loader=FileSystemLoader(_HTML_DIR))
             telemetry_content = env.get_template("telemetry.html").render(
                 heap_app_id=heap_app_id, heap_user_identity=heap_user_identity
             )
-            html_content = html_content.replace("</head>", telemetry_content)
+            injected_head_content.append(telemetry_content)
+
+        if autoreload:
+            autoreload_content = env.get_template("autoreload.html").render(
+                etag=app_etag
+            )
+            injected_head_content.append(autoreload_content)
+
+        injected_head_content.append("</head>")
+        html_content = html_content.replace("</head>", "\n".join(injected_head_content))
         return HTMLResponse(html_content)
 
+    @app.get("/api/reload")
+    async def reload(request: Request):
+        if "If-None-Match" not in request.headers:
+            raise HTTPException(
+                status_code=400,
+                detail="Request to reload endpoint must have an If-None-Match header",
+            )
+        current_client_etag = request.headers["If-None-Match"]
+        if app_etag == current_client_etag:
+            return Response(status_code=304)
+
+        return Response()
```

Comment:
~Is the default status code 200 (for a `Response()`)?~ Nevermind the tests have answered my question.